### ISSUE TITLE
no-unreachable: shouldn't be triggered by a semicolon after switch

### ISSUE THREAD

**Issue by himdel** [eslint/eslint#6295]
**What version of ESLint are you using?**
v2.11.0

**What parser (default, Babel-ESLint, etc.) are you using?**
Occurs with either espree or babel-eslint.

**Please show your full configuration:**
Anything with `no-unreachable` enabled..

**What did you do? Please include the actual source code causing the issue.**

This triggers  `5:4  error  Unreachable code  no-unreachable`

```js
function foo(bar) {
  switch (bar) {
    default:
      return bar;
  };
}
```

This doesn't..

```js
function foo(bar) {
  switch (bar) {
    default:
      return bar;
  }
}
```

(the only difference being in the semicolon after the switch)

**What did you expect to happen?**

I would have expected `no-unreachable` **not** to trigger on semicolon.. (there's other rules for that :)).

**What actually happened? Please include the actual, raw output from ESLint.**

`5:4  error  Unreachable code  no-unreachable`

**Comment by michaelficarra** [comment]
That semicolon is a separate statement following the switch. Since you're returning from all cases in the switch, execution will never reach there. The rule is working as intended.

**Comment by mysticatea** [comment]
Yes, @michaelficarra is correct. An empty statement exists after the switch statement if a semicolon exists. Then `no-unreachable` rule warns the empty statement.
This is similar to:

```js
function foo() {
    return;
    ; // Unreachable code
}
```

However, as @himdel mentioned, `no-unreachable` may not need to warn empty statements because we have `no-extra-semi` rule for empty statements. Thought?

**Comment by platinumazure** [comment]
@mysticatea Isn't it true that someone could turn off `no-extra-semi` though? Then they would get no warning at all, if EmptyStatements are allowed by `no-unreachable`.

Personally, I see nothing wrong with having two lint errors show up for two different issues. The unreachable-ness of the semicolon has nothing to do with the fact that it's an unnecessary/empty statement; the emptiness of the statement has nothing to do with the fact that it is unreachable.

Personally I would say this is working as designed and we don't need to enhance it.

**Comment by mysticatea** [comment]
> they would get no warning at all, if EmptyStatements are allowed by no-unreachable.

Exactly it's what I said. This issue's request sounded like "I want to use extra semi, but `no-unreachable` didn't allow it." to me.

**Comment by himdel** [comment]
Exactly, I do understand that from the parser's standpoint it really is an unreachable statement. But it means there's currently no way to allow semicolon after switch and get no warnings while still wanting to be warned about nonempty unreachable statements.. 

Maybe just adding an `ignoreEmpty` option to `no-unreachable` would work? (and yes, I'm willing to implement that :-))

**Comment by platinumazure** [comment]
I'm not sure I understand why you think you need the semicolon after switch
in the first place. It's completely unnecessary and it's not allowed by the
language (that is, a SwitchStatement in the grammar does not allow an
optional semicolon like simple statements do, and any semicolon you do
place is unambiguously an empty and unnecessary statement).

**Comment by himdel** [comment]
> it's not allowed by the language

Well it's certainly not recognized by the language as part of that switch, and I agree that it makes about as little sense to add a semicolon after a swtich, as after an `if`, or a function declaration. However, there is nothing in the language that would make such a use incorrect .. and the simple truth is, some programmers simply prefer it because it feels more consistent, and believe me, none of them thing of that semicolon as introducing an empty statement, no matter what the language spec says.

So.. I guess the real question is .. does `eslint` see itself as a project setting a policy, or a project providing the means to validate against custom policices? (as in, ["mechanism, not policy"](https://en.wikipedia.org/wiki/Separation_of_mechanism_and_policy)).

And to be clear, I'm definitely not proposing such an option to be enabled by the `recommended` rules, I just think that there should be a way for projects which really want it to allow such use, without having to use their own fork of `no-unreachable`.

So from my point of view, enforcing no semicolon after switch is the job of `no-semi`, not `no-unreachable` .. and there should be a way to not enforce `no-semi` but still be able to use `no-unreachable`..

**Comment by himdel** [comment]
Maybe a specific use case..

Consider `{ "no-semi": "warn", "no-unreachable": "error" }` .. sounds reasonable, right?

Well, without something like the proposed `ignoreEmpty`, a simple unreachable statement will trigger 1 error, whereas an almost invisible comma after switch will trigger 1 warning and 1 error, and I think we can all agree that anybody who uses this combination would expect just 1 warning in this case.

And.. if you contribute code somewhere and are not aware of the rules, the error about unreachable code will simply feel nonsensical, there is no actual code, just that almost invisible semicolon...

**Comment by ilyavolodin** [comment]
@himdel Sad truth is, there's no way we can support every possible scenario. We have to support cases that are widely used, and provide options for things that are common. I have not seen this style before, nor am I aware of any style guides that use this style. I think you will have to either disable this rule in general, or use inline comments to disable it around switch cases, or close the rule and release your own plugin with modified version.

**Comment by himdel** [comment]
@ilyavolodin OK, understood .. if we end up need this, I'll do a separate module, closing the issue, thanks anyway :)
